File: Mechanica Engine/src/renderers/WorldShader.java
package renderers;

import org.lwjgl.util.vector.Matrix4f;

/**
 * Created by domin on 15 Apr 2017.
 */
public abstract class WorldShader extends ShaderProgram {

    protected int transformationMatrix;
    protected int projectionMatrix;
    protected int viewMatrix;

    public WorldShader(String vertexFile, String fragmentFile) {
        super(vertexFile, fragmentFile);
    }

    public void getMatrixLocations(){
        transformationMatrix = super.getUniformLocation("transformationMatrix");
        projectionMatrix = super.getUniformLocation("projectionMatrix");
        viewMatrix = super.getUniformLocation("viewMatrix");
    }

    protected static String getPath(){
        return WorldShader.class.getProtectionDomain().getCodeSource().getLocation().getPath();
    }

    public void loadTransformationMatrix(Matrix4f matrix){
        super.loadMatrix(transformationMatrix, matrix);
    }

    public void loadProjectionMatrix(Matrix4f projection){
        super.loadMatrix(projectionMatrix, projection);
    }

    public void loadViewMatrix(Matrix4f view){
        super.loadMatrix(viewMatrix, view);
    }


}

File: lwjgl3-test/src/renderer/WorldShader.kt
package renderer

import org.joml.Matrix4f

/**
 * Created by domin on 15 Apr 2017.
 */
abstract class WorldShader(vertexFile: String, fragmentFile: String) : ShaderProgram(vertexFile, fragmentFile) {

    protected var transformationMatrix: Int = 0
    protected var projectionMatrix: Int = 0
    protected var viewMatrix: Int = 0

    fun getMatrixLocations() {
        transformationMatrix = super.getUniformLocation("transformationMatrix")
        projectionMatrix = super.getUniformLocation("projectionMatrix")
        viewMatrix = super.getUniformLocation("viewMatrix")
    }

    fun loadTransformationMatrix(matrix: Matrix4f) {
        super.loadMatrix(transformationMatrix, matrix)
    }

    fun loadProjectionMatrix(projection: Matrix4f) {
        super.loadMatrix(projectionMatrix, projection)
    }

    fun loadViewMatrix(view: Matrix4f) {
        super.loadMatrix(viewMatrix, view)
    }

    companion object {

        protected val path: String
            get() = WorldShader::class.java.protectionDomain.codeSource.location.path
    }


}
